### ISSUE TITLE
[CLIFF] Training details

### ISSUE THREAD

**Issue by pangyyyyy** [huawei-noah/noah-research#101]
Hi authors, thanks so much for the great work!! 

Following up from #94, I am trying to train CLIFF on Human3.6M, MPI-INF-3DHP, COCO (your pseudo GT), MPII (your pseudo GT) and 3DPW-train set with partitions 0.4, 0.1, 0.2, 0.1, 0.2 respectively. I am using 224x224 images with lr of 1×e−4 and Adam Optimizer following the details listed in the paper. However, I was unable to reach the reported accuracy (current PA-MPJPE >50) using HMR pipeline with the addition of `bbox_info` generated from the ground-truth annotations.

I would like to get your advice on how to improve the accuracy. I have a few queries:
- Do you plan to release the training code soon? 
- Is this trained using a set-up similar to HMR (regression-based) or SPIN (optimization-based)? 
- Did you initialise with any pretrained models or was it trained from scratch?
- Is the `bbox_info` used during training generated from ground-truth annotations?
- You mentioned that CLIFF was trained `for 244K steps with batch size 256`. Did you use distributed training? If so, how many GPUs and what set up was used?
- While the paper mentioned that the `cropped images are resized to 224x224` during training, the reported accuracy is using the model trained on 256x192 images. I was wondering if you have the reported accuracy using 224x224 images?
- You used a variety of augmentations - `Data augmentation includes random rotations and scaling, horizontal flipping, synthetic occlusion [45], and random cropping [15,44]`. What is the augmentation probability and scale i.e. flip probability of 0.5? size of occlusion?

Thanks!

**Comment by zhihaolee** [comment]
Had you used the CLIFF supervision via the perspective projection in the full frames? Please make sure that you do it correctly by visualization?

1. No, please refer to #96 .
2. Trained using a set-up similar to HMR (regression-based), without optimization.
3. Initialized with the ImageNet-pretrained weights.
4. Yes, `bbox_info` is generated from the ground-truth annotation during training.
5. Yes, we use the distributed training on 4 Tesla V100. For training on a single low-memory GPU, I think the similar performance could be achieved by training 4x more steps with batch size of 64.
6. I lost the result data of 224x224, but they are similar to 256x192 as I can remember. It would be great if you can do an ablation study to check it .
7. The augmentation setting is the same with SPIN and PARE. We use the random cropping only for the AGORA evaluation.

**Comment by pangyyyyy** [comment]
@zhihaolee Thanks for the clarification!

1. What do you mean by CLIFF supervision via perspective projection in full frames? Do you mean the supervision of keypoints 2d (perspective projection was done for kp2d in SPIN)? https://github.com/nkolot/SPIN/blob/5c796852ca7ca7373e104e8489aa5864323fbf84/train/trainer.py#L170-L174

2.  With a batch size of 256 on 4GPUs, training for 224K train steps is equivalent to training for 732 epochs (1 epoch contains 306 train steps/ 320K samples).  The normal SPIN/HMR pipeline trains for only 100 epochs. Just wanted to check, what was your total training time?

Thanks!

**Comment by zhihaolee** [comment]
1. That's the one in the cropped images. Please read again the second part of Section 3.3 in the paper which is about the reprojection loss.
2. 5 days for HRNet-W48, half time if ResNet-50. By the way, we use SyncBN with batch_size 64 on each GPU, 4 GPUs in total.

**Comment by pangyyyyy** [comment]
@zhihaolee Thanks so much for the clarification! I was previously doing supervision in cropped frames, that might be the cause. From what I understand in eq. 10 section 3.3 in the paper, only `L_2D^full` 2D keypoint supervision is in full frames? 

I am surprised that there's such a big difference between cropped/ full frame supervision. I can't seem to find the ablation in the paper, did you have any experiments on this?

**Comment by zhihaolee** [comment]
1. Yes, only `L_2D^full` 2D keypoint supervision is in full frames.
2. No, no ablation study of CS on 3DPW.
3. Just minusing no normalization in our implemetation. Other implementations are okay.
4. The core of CS is given here:
https://github.com/huawei-noah/noah-research/blob/ce43ed04b68fa33e3d521cadcc4986b492ead9e4/CLIFF/demo.py#L132-L134

**Comment by pangyyyyy** [comment]
@zhihaolee got it, thanks for your helpful reply!!

A few last questions:
1. What is the loss i.e. MSE/L1 and loss weight you used for supervision of L_smpl, L_3D, L2D_full? SPIN and HMR uses MSE loss, From eq 9 in the paper, it seems like keypoints 2D loss is MAE loss? 
2. What is the point of minusing off full image center? From eq 9, it seems that minusing off the full image center for both will cancel out i.e. (gt_kp2d-center) - (pred_kp2d - center) = gt_kp2d - pred_kp2d
3. Without normalization, wouldn't the loss for keypoints2d be really large since the errors are in full image space? Did you reduce the loss weight for keypoints 2d supervision? 

Thanks!

**Comment by zhihaolee** [comment]
1. The same as SPIN.
2. Yes, they are the same. But in our implematation, the `pred_kp2d ` from perspective projection is already aligned at the full image center.
3. You are right, sorry for my mistake. We normalize the `keypoints2d ` by the bounding box size before calculating the loss.

**Comment by pangyyyyy** [comment]
@zhihaolee Thanks for the clarification!!

**Comment by pangyyyyy** [comment]
@zhihaolee I loaded the pretrained checkpoints and inspect the epoch they are saved from: 

'data/checkpoints/res50-PA45.7_MJE72.0_MVE85.3_3dpw.pt' : epoch 5/step 6095
'data/checkpoints/hr48-PA43.0_MJE69.0_MVE81.2_3dpw.pt' : epoch 7/step 8533 

The paper said to train for 224K steps i.e. about 200 epochs. However, it seems that the provided checkpoints/ best models were saved from epoch 5 and 7, is this true? 

How did it achieve convergence so quickly? and do you mind sharing the training log if possible?

Thanks!

**Comment by cians** [comment]
@pangyyyyy 
Hello，Have you reproduced the results of the paper so far？ I would like to know the details during the training.
Thanks!

**Comment by LightOfMonet** [comment]
Can you get to this SOTA number of PA43.0 with these optimizations ?  
Does the test code use SPIN？

**Comment by wulalah** [comment]
> 1. The same as SPIN.
> 2. Yes, they are the same. But in our implematation, the `pred_kp2d` from perspective projection is already aligned at the full image center.
> 3. You are right, sorry for my mistake. We normalize the `keypoints2d` by the bounding box size before calculating the loss.

Dear author, you mentioned the normalized factor is the bounding box size in the third answer. Does the bounding box size refer to scale*200 or image side resolution 224?
